etl/parse_email_save_attachment/parse_messages.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+
+import os
+import pytz
+import boto3
+import email
+import tempfile
+from email import policy
+from operator import itemgetter
+from datetime import datetime
+
+
+def get_most_recent_file(bucket_name, prefix):
+    s3 = boto3.client("s3")
+
+    # List objects within the specified bucket and prefix
+    response = s3.list_objects_v2(Bucket=bucket_name, Prefix=prefix)
+
+    # Check if the bucket is empty or the prefix doesn't match any files
+    if "Contents" not in response:
+        return "No files found in the bucket with the specified prefix."
+
+    # Sort the files by last modified date
+    files = sorted(response["Contents"], key=itemgetter("LastModified"), reverse=True)
+
+    # Get the most recent file
+    most_recent_file = files[0]
+    return most_recent_file["Key"]
+
+
+def get_file_content(bucket_name, file_key):
+    s3 = boto3.client("s3")
+
+    # Get the object
+    obj = s3.get_object(Bucket=bucket_name, Key=file_key)
+
+    # Get the object's content
+    content = obj["Body"].read().decode("utf-8")
+    return content
+
+
+def parse_email_from_s3(email_content):
+    # Decode the string into bytes
+    email_content_bytes = email_content.encode("utf-8")
+    parsed_email = email.message_from_bytes(email_content_bytes, policy=policy.default)
+
+    # Create a temporary directory to store attachments
+    temp_dir = tempfile.mkdtemp()
+    if parsed_email.is_multipart():
+        # If email has multiple parts (like attachments or different content types)
+        for part in parsed_email.iter_parts():
+            if part.is_attachment():
+                # Save the attachment in the temporary directory
+                filename = part.get_filename()
+                if filename:
+                    filepath = os.path.join(temp_dir, filename)
+                    with open(filepath, "wb") as f:
+                        f.write(part.get_payload(decode=True))
+
+    return temp_dir
+
+
+import glob
+
+
+def upload_attachments_to_s3(bucket, prefix, location):
+    s3 = boto3.resource("s3")
+
+    # Create a new folder with the current date and time in Central Time
+    central = pytz.timezone("US/Central")
+    folder_name = datetime.now(central).strftime("%Y%m%d-%H%M%S")
```

Comment:
Would it be possible to use the email's received date for the name of these archived folders so if we stopped getting emails for some reason it wouldn't appear that we're still getting new data? I'm not sure how often this will run and how often we're getting emails but it should also reduce the number of duplicate entries.

Reply:
i always get nervous when i see a non-UTC timestamp without a timezone offset. could we get by on UTC here instead of fussing with timezones?

Reply:
I have the same question. Here is what I'm seeing in S3 when running this several times in a row - all these contain the same data. I could be missing a detail about the process, but I like Charlie's idea to create folders based off a timestamp tied to what has been received so the ETL would know to skip the upload if there isn't new data.
<img width="240" alt="Screenshot 2024-02-28 at 11 19 07 AM" src="https://github.com/cityofaustin/dts-maximo-geo-integration/assets/37249039/58f1cf62-dcf5-48c2-be4f-f5dedf744311">

Reply:
Great questions y'all, let me take them one by one:

## Charlie's question

@Charlie-Henry asks if we can use the email's received date. I'm not sure we can -- the team opted in a recent reining meeting to not parse these emails upon receipt, but rather to reuse the VZ pattern and parse them via an ETL. As a result, it's possible and likely that we will end up parsing the same email multiple times. If the email's sent timestamp was a unique identifier of an email, then we'd be able to use it, but otherwise, I don't think we can ensure that we don't receive two emails at the same time. I do realize that it's very unlikely..

That said, I totally do recognize the value in being able to see something that indicates we're not processing new information and just churning over the same email over and over, so I am going to extend the program to include a snippet of the sha256 of the raw email we received. That will end up not changing from time to time when this runs on the same input, so we'll have an indication in the file listing of what's happening. I think this will serve the spirit of the request, just not using the timestamp.

## John's concern

@johnclary recognizes the challenges of leaving a timestamp without an indication of what timezone it is. He's totally right, and I am going to adjust the code to use UTC, and further, it will include an indication in the directory name that the timestamp is UTC. :100: 

## Mike's question

@mddilley shares the same question as Charlie and he observes that we are getting the same output from multiple runs. Based on how we scoped this, this is by design - we'll get the same output until a new email is received from our upstream source. However, Mike offers a great idea based on what Charlie suggested -- using a finger print to know we've already processed some data. In lieu of the timestamp found in the email, I am going to hash the whole email payload, and store that in S3 along with the extracted attachments. Using that, I can enhance the program to escape early if it determines that it's already seen certain data.


Great suggestions and questions - thanks y'all!

Reply:
This is following the same pattern as the EMS/AFD/VZ email processing, does the the VZ setup also potentially process the same email multiple times?

Reply:
> This is following the same pattern as the EMS/AFD/VZ email processing, does the the VZ setup also potentially process the same email multiple times?

I believe it would yes, but we didn't give it a ton of thought because for those, we know the interval that we're supposed to receive the data, so we only run it once a day. If it does build up cruft, it does it a slower rate than this one would if geo were to ask for it to run real frequently.

Reply:
Cool - y'all's suggestions and concerns here were :100: :dart:. Thank you so much for them. Here's what's changed:

* Use UTC not Central and note it: 16178bab9c0718bec2caa90375bfaea6eb9d020d
* Validate 4 headers to increase our confidence that the sender is who they say they are: c8c6328091478d7f132f9b480743d6fe3b5c6114
* Use sha256 to encode some state in the directories of files processed. Exit early if we've seen an email before. fb413246cec36df3fe9b82b7e3e9cdffccd447d0 & 5ca2a3b98eac9df501814f6a96aea69ee489fc6d

Thanks again, and let me know if I can do anything else on these :pray: